### PR TITLE
install gitea

### PR DESCRIPTION
**Pull request by nimakaviani**
fixes 32

### CODE REVIEW

**Comment by cmoulliard** [comment]
Why do you install gitea as kubernetes resources and not using a Helm chart instead as I described in my issue #32 ?

File: pkg/controllers/localbuild/resources/gitea/values.yaml
Comment (cmoulliard) [inline pkg/controllers/localbuild/resources/gitea/values.yaml:1]: This values file is only needed if you install gitea as a helm chart (and this is what I recommend to do) as it allows to externalize the configuration/parameters. 

If you deploy it using the helm chart, then additional parameters will be needed to expose the gitea server as UI on the cluster, etc as documented here #32

**Comment by nimakaviani** [comment]
The values file was used in `helm template` to generate and put the vanilla k8s resources under `resources/gitea/k8s`. I kept it there as I am still experimenting with what resources we want to configure (including the ingress and service information you mentioned in #32). we can remove it once we finalize the PR,

If we go down the route of installing gitea using the helm chart, we will have to either pull the helm go libraries in or create a dependency to helm, which I'd like to avoid. 

As we discussed in #5 handling of any package or tool outside the initial bootstrapping should be handed over to Argo Apps, which can then take care of helm or kustomize or any other package types. That's why I am hesitant to introduce helm as part of the bootstrapping mechanism. its a draft PR, so open to suggestions here.

**Comment by cmoulliard** [comment]
> If we go down the route of installing gitea using the helm chart, we will have to either pull the helm go libraries in or create a dependency to helm, which I'd like to avoid.

Why ? You are also today creating a hard dependency today as you must maintain the ArgoCD YAML file able to install argocd as core component + controller tracking its installation while using helm go lib we could more easily install the core components (argo cd, vault, argo workflow, cert manager) before to delegate to ArgoCD the job to install the `pluggable` Application CRs = Components

**Comment by cmoulliard** [comment]
> we can remove it once we finalize the PR,

Ok.

**Comment by nabuskey** [comment]
@nimakaviani Is this going to be ready for review soon? I would like to get Gittea integration going.

**Comment by nabuskey** [comment]
We should reference the secret for the admin user credentials instead of directly using it in the pod template. See: https://gitea.com/gitea/helm-chart#admin-user

This way other controllers can reference it and get the credentials to talk to Gitea api. This does mean we have to helm template then add another manifest for the secret, which is a bit annoying and may cause confusions among users. So we should document and warn.

**Comment by nabuskey** [comment]
For helm values, we will also need to specify endpoint domain and URLs. Otherwise, API responses from Gitea for anything referencing base url has the default values of `git.example.com`. 

For my local testing, I am using this, then port-forward. 

```yaml
gitea:
  admin:
    existingSecret: gitea-admin-secret
  config:
    database:
      DB_TYPE: sqlite3
    session:
      PROVIDER: memory
    cache:
      ADAPTER: memory
    queue:
      TYPE: level
    server:
      DOMAIN: localtest.me
      ROOT_URL: 'http://localtest.me:3000'
```

**Review by jessesanford — APPROVED** [review]
Love it. Thanks for drying up those local installs.

File: pkg/controllers/localbuild/argo.go [hunk 1]
```diff
@@ -5,105 +5,47 @@ import (
...
-	}
-
-	return k8s.ConvertRawResourcesToObjects(scheme, rawResources)
-}
-
-func newArgoNamespace() *corev1.Namespace {
-	return &corev1.Namespace{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "argocd",
+func (r LocalbuildReconciler) ReconcileArgo(ctx context.Context, req ctrl.Request, resource *v1alpha1.Localbuild) (ctrl.Result, error) {
```
Comment (nabuskey) [inline pkg/controllers/localbuild/argo.go:19]: Let's use pointer receiver. I know we don't need pointers for these but rest of them do. We are currently mixing value and pointer receivers.

File: pkg/controllers/localbuild/nginx.go [hunk 2]
```diff
@@ -3,133 +3,38 @@ package localbuild
...
-
-func newNginxNamespace() *corev1.Namespace {
-	return &corev1.Namespace{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: nginxNamespace,
-		},
-	}
-}
 
 func (r LocalbuildReconciler) ReconcileNginx(ctx context.Context, req ctrl.Request, resource *v1alpha1.Localbuild) (ctrl.Result, error) {
```
Comment (nabuskey) [inline pkg/controllers/localbuild/nginx.go:19]: Same thing here. Let's use pointer receiver.

File: pkg/controllers/localbuild/gitea.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+const (
+	giteaNamespace = "gitea"
+)
+
+//go:embed resources/gitea/k8s/*
+var installGiteaFS embed.FS
+
+func (r LocalbuildReconciler) ReconcileGitea(ctx context.Context, req ctrl.Request, resource *v1alpha1.Localbuild) (ctrl.Result, error) {
```
Comment (nabuskey) [inline pkg/controllers/localbuild/gitea.go:19]: Same thing here. Let's use pointer receiver.

File: pkg/controllers/localbuild/nginx.go [hunk 4]
```diff
@@ -3,133 +3,38 @@ package localbuild
...
-	nginxNamespace  string = "ingress-nginx"
-	nginxDeployment string = "ingress-nginx-controller"
+	nginxNamespace string = "ingress-nginx"
 )
 
 //go:embed resources/nginx/k8s/*
 var installNginxFS embed.FS
-var timeout = time.After(3 * time.Minute)
-
-func RawNginxInstallResources() ([][]byte, error) {
```
Comment (nabuskey) [inline pkg/controllers/localbuild/nginx.go:32]: Can we keep the ability to get embedded resources outside of these packages? Keep the functions or export the variable. The git repository reconciler needs a way to retrieve these to sync with gittea repos. You don't have to pass local dir ref and we have a distinction between repositories created with embedded contents vs local directory. 

```yaml
apiVersion: idpbuilder.cnoe.io/v1alpha1
kind: GitRepository
metadata:
  name: argocd
  namespace: default
spec:
  giteaURL: "http://localhost:3000"
  source:
    embeddedAppName: "argocd"
    type: embedded
```

File: api/v1alpha1/localbuild_types.go [hunk 5]
```diff
@@ -20,7 +20,14 @@ type EmbeddedArgoApplicationsPackageConfigSpec struct {
 	Enabled bool `json:"enabled,omitempty"`
 }
 
+// GitConfigSpec controls what git server to use for the idpbuilder
+// It can take on the values of either gitea or gitserver
+type GitConfigSpec struct {
+	Type string `json:"type,omitempty"`
```
Comment (nabuskey) [inline api/v1alpha1/localbuild_types.go:26]: Wish there was a way to put the reference to secrets that contain gitea admin credentials in Spec or Status. I think for this implementation we can just document secret requirements. Not a blocker imo, but we need to open an issue.

File: pkg/controllers/localbuild/installer.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+type EmbeddedInstallation struct {
+	name         string
+	resourcePath string
+	namespace    string
+
+	// skips waiting on expected resources to become ready
+	skipReadinessCheck bool
+
+	// name and gvk pair for resources that need to be monitored
+	expectedResources map[string]schema.GroupVersionKind
```
Comment (nabuskey) [inline pkg/controllers/localbuild/installer.go:37]: I think the name of this field should be `monitoredResources`. I first thought this was for making sure resources specified are present during installation.

File: api/v1alpha1/localbuild_types.go (hunk 5 above)
Comment (nimakaviani) [reply api/v1alpha1/localbuild_types.go:26]: well the secret name is hardcoded here, so maybe we can just use it for now. i like the idea https://github.com/cnoe-io/idpbuilder/pull/39/files#diff-6eb0197d5d12e1ca681eac36170032f5d01e997ea21bdbd87803a302f9df254fR582-R590

File: pkg/controllers/localbuild/argo.go (hunk 1 above)
Comment (nimakaviani) [reply pkg/controllers/localbuild/argo.go:19]: good call. not sure how it fell through

File: pkg/controllers/localbuild/nginx.go (hunk 4 above)
Comment (nimakaviani) [reply pkg/controllers/localbuild/nginx.go:32]: lets revisit it as part of pushing the files to the repos in the next PR. We will make adjustments where needed.

**Pull request merged**